### DISCUSSION TITLE
NTFS Alternate Streams in .NET

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/runtime#62360] [Q&A] [answered]
If I run the following in CMD, I can write content to an alternate data steam and then list the streams:
```
echo "Hello" > Test.txt:secret

dir /R test.txt
```
Output:
```
 Volume in drive C has no label.
 Volume Serial Number is 7042-D090

 Directory of C:\Users\TonyV

12/03/2021  03:19 PM                 0 Test.txt
                                    10 Test.txt:secret:$DATA
               1 File(s)              0 bytes
               0 Dir(s)   6,258,278,400 bytes free
```

If I want to do it in .NET, what do I do?
```
System.IO.File.WriteAllText("Test.txt:secret", "Hello");
//What is the second line?
```

**Comment by huoyaoyuan** [comment, marked as answer]
@dotnet/area-system-io I believe it's not supported yet, and can be a feature request.

**Reply by danmoseley** [reply]
Right, @TonyValenti since there is no existing API, you could open an issue in this repo to propose a shape. (I do not see an existing issue, but I see some discussion in [this one](https://github.com/dotnet/runtime/issues/49604))

It is not just the API that would need thinking through, but also research and propose the behavior it would have on other OS/file systems. It is unlikely we would want to add a Windows-only API.

**Comment by adamsitnik** [comment]
Do you mean how you can print directory listing using .NET?

You could use [FileSystemEnumerable](https://docs.microsoft.com/en-us/dotnet/api/system.io.enumeration.filesystemenumerable-1?view=net-6.0#examples) to iterate over the files and just print the information using `Console.WriteLine`

**Reply by TonyValenti** [reply]
@adamsitnik -
The DIR command listed in my example will list alternate data streams for a file.

I Was curious because, based on the format of the DIR command, I assume that there is some api that normally returns contents of folders but will also return a list of streams.

**Reply by msedi** [reply]
@TonyValenti: Does this help: https://stackoverflow.com/questions/604960/how-to-read-and-modify-ntfs-alternate-data-streams-using-net.

I played a while ago with it because I wanted to started a large set of metadata in the ADF. But then I moved back, because I didn't want to bind my code to NTFS.
